File: 01-游戏开发/04-游戏引擎/Godot/节点生命周期/go/nodelifecycle.go
// nodelifecycle.go — Godot 节点生命周期与帧处理顺序的最小模型（与 python/nodelifecycle.py 同题）。
//
// 事实来源：Godot 官方 class_node / nodes_and_scene_instances / idle_and_physics_processing
// 三篇文档。要点：_enter_tree 父先于子；_ready 子先于父（逆序，且 @onready 在其之前）；
// queue_free 延迟到当前帧处理结束后并连带释放子节点；_process 随帧率而
// _physics_process 固定 60 次/秒，两者不同步。
package main

import (
	"fmt"
	"strings"
)

const (
	physicsFPS = 60
	stepUS     = 1000000 / physicsFPS // 16.666 ms
)

type node struct {
	name     string
	world    *world
	parent   *node
	children []*node
	inTree   bool
	ready    bool
	alive    bool
	process  bool
	physics  bool
	pCalls   int
	yCalls   int
	deltas   []float64
}

func newNode(name string, w *world, physics bool) *node {
	return &node{name: name, world: w, alive: true, process: true, physics: physics}
}

type world struct {
	root         *node
	log          []string
	pendingFree  []*node
	physicsAcc   int
}

func newWorld() *world {
	w := &world{}
	w.root = newNode("root", w, false)
	w.root.inTree = true
	w.root.ready = true
	return w
}

func (w *world) enterTree(n *node) { // 自顶向下：父先于子
	n.inTree = true
	w.log = append(w.log, n.name+":enter_tree")
	for _, c := range append([]*node{}, n.children...) {
		w.enterTree(c)
	}
}

func (w *world) markReady(n *node) { // 自底向上：子先于父
	for _, c := range append([]*node{}, n.children...) {
		w.markReady(c)
	}
	w.log = append(w.log, n.name+":onready") // @onready 在 _ready 之前
	n.ready = true
	w.log = append(w.log, n.name+":ready")
}

func (w *world) exitTree(n *node) {
	n.inTree = false
	w.log = append(w.log, n.name+":exit_tree")
	for _, c := range append([]*node{}, n.children...) {
		w.exitTree(c)
	}
}

func (n *node) addChild(c *node) {
	if c.parent != nil {
		c.parent.removeChild(c)
	}
	c.parent = n
	n.children = append(n.children, c)
	if n.inTree {
		n.world.enterTree(c)
		n.world.markReady(c)
	}
}

func (n *node) removeChild(c *node) {
	for i, x := range n.children {
		if x == c {
			n.children = append(n.children[:i], n.children[i+1:]...)
			break
		}
	}
	c.parent = nil
	if c.inTree {
		n.world.exitTree(c)
	}
}

func (n *node) queueFree() { n.world.pendingFree = append(n.world.pendingFree, n) }

func (w *world) freeNode(n *node) {
	if !n.alive {
		return
	}
	if n.parent != nil {
		n.parent.removeChild(n)
	}
	if n.inTree {
		w.exitTree(n)
	}
	for _, c := range append([]*node{}, n.children...) { // 连带释放所有子节点
		w.freeNode(c)
	}
	n.children = nil
	n.alive = false
	w.log = append(w.log, n.name+":free")
}

func (w *world) all(n *node) []*node {
	out := []*node{n}
	for _, c := range n.children {
		out = append(out, w.all(c)...)
	}
	return out
}

// frame 推进一帧：先按固定步长跑物理，再跑一次 idle，最后处理 queue_free。
func (w *world) frame(dtUS int) int {
	w.physicsAcc += dtUS
	steps := 0
	for w.physicsAcc >= stepUS {
		w.physicsAcc -= stepUS
		steps++
		for _, n := range w.all(w.root) {
			if n.alive && n.inTree && n.physics {
				n.yCalls++
			}
		}
	}
	for _, n := range w.all(w.root) {
		if n.alive && n.inTree && n.process {
			n.pCalls++
			n.deltas = append(n.deltas, float64(dtUS)/1e6)
		}
	}
	for _, n := range append([]*node{}, w.pendingFree...) {
		w.freeNode(n)
	}
	w.pendingFree = nil
	return steps
}

func pacing(totalUS, frameUS int) []int {
	out := []int{}
	for left := totalUS; left > 0; {
		s := frameUS
		if s > left {
			s = left
		}
		out = append(out, s)
		left -= s
	}
	return out
}

func main() {
	w := newWorld()
	parent := newNode("parent", w, false)
	child := newNode("child", w, false)
	grand := newNode("grand", w, false)
	child.addChild(grand)
	parent.addChild(child)
	w.root.addChild(parent)
	fmt.Printf("回调顺序 = %s\n", strings.Join(w.log, " → "))

	w2 := newWorld()
	detached := newNode("detached", w2, false)
	kid := newNode("kid", w2, false)
	detached.addChild(kid)
	fmt.Printf("挂到游离父节点时回调数 = %d（应为 0）\n", len(w2.log))
	w2.root.addChild(detached)
	fmt.Printf("补挂后 = %s\n", strings.Join(w2.log, " → "))

	w3 := newWorld()
	a := newNode("a", w3, true)
	ac := newNode("a_child", w3, false)
	a.addChild(ac)
	w3.root.addChild(a)
	a.queueFree()
	w3.frame(stepUS)
	fmt.Printf("queue_free：本帧 _process=%d，帧末 alive=%v，子节点 alive=%v\n",
		a.pCalls, a.alive, ac.alive)

	w4 := newWorld()
	d := newNode("d", w4, true)
	w4.root.addChild(d)
	for _, dt := range pacing(1000000, 16666) {
		w4.frame(dt)
	}
	w5 := newWorld()
	e := newNode("e", w5, true)
	w5.root.addChild(e)
	for _, dt := range pacing(1000000, 33333) {
		w5.frame(dt)
	}
	fmt.Printf("1 秒：60fps → process=%d physics=%d；30fps → process=%d physics=%d\n",
		d.pCalls, d.yCalls, e.pCalls, e.yCalls)

	w6 := newWorld()
	g := newNode("g", w6, true)
	w6.root.addChild(g)
	s1 := w6.frame(1000) // 1ms：不足一个物理步
	s2 := w6.frame(33333)
	fmt.Printf("短帧物理步=%d（process 仍为 1），长帧物理步=%d\n", s1, s2)
}

File: 01-游戏开发/04-游戏引擎/Godot/节点生命周期/python/nodelifecycle.py
"""nodelifecycle.py — Godot 节点生命周期与场景/帧处理顺序的最小模型与自检.

事实来源（本机 curl 落地后实读的 Godot 官方文档）：
  * docs.godotengine.org/en/stable/classes/class_node.html
      —— 节点加入 SceneTree 时先 NOTIFICATION_ENTER_TREE / _enter_tree()，
         **父节点先于子节点**；全部加入后收到 NOTIFICATION_READY / _ready()，
         对一组节点 **_ready() 从子节点开始逆序到父节点**；
         完整顺序：父 _enter_tree → 子 _enter_tree → 子 _ready → 父 _ready（递归）；
         Object.free() 或 queue_free() 都会连带释放它的所有子节点。
  * docs.godotengine.org/en/stable/tutorials/scripting/nodes_and_scene_instances.html
      —— queue_free() 把节点排到**当前帧处理结束之后**删除；free() 立即销毁，
         已有引用会立刻变 null，官方推荐 queue_free()；
         实例化分两步：load(PackedScene) 得到资源 → PackedScene.instantiate() 得到节点树；
         @onready 成员在 _ready() **之前**初始化；get_node 支持 ".." 与 "/root" 绝对路径。
  * docs.godotengine.org/en/stable/tutorials/scripting/idle_and_physics_processing.html
      —— _process(delta) 的频率取决于帧率，delta = 距上次调用经过的秒数；
         _physics_process(delta) 按固定间隔跑（Physics Fps 默认 60 次/秒）；
         _process 与物理**不同步**；可用 Node.set_process() 开关。
"""

from __future__ import annotations

from typing import List, Optional

PHYSICS_FPS = 60
STEP_US = 1_000_000 // PHYSICS_FPS        # 16.666 ms

_ASSERTIONS = 0


def check(label: str, cond: bool, detail: str = "") -> None:
    global _ASSERTIONS
    if not cond:
        raise AssertionError(f"{label} 失败: {detail}")
    _ASSERTIONS += 1
    print(f"ok {_ASSERTIONS:>2} {label}" + (f"  [{detail}]" if detail else ""))


class Node:
    def __init__(self, name: str, world: "World", physics: bool = False) -> None:
        self.name = name
        self.world = world
        self.parent: Optional["Node"] = None
        self.children: List["Node"] = []
        self.in_tree = False
        self.ready = False
        self.alive = True
        self.process_on = True
        self.physics_on = physics
        self.process_calls = 0
        self.physics_calls = 0
        self.deltas: List[float] = []
        self.physics_deltas: List[float] = []

    # ---- 树操作 ----------------------------------------------------
    def add_child(self, child: "Node") -> None:
        if child.parent is not None:
            child.parent.children.remove(child)
        child.parent = self
        self.children.append(child)
        if self.in_tree:                       # 父已在树中 ⇒ 立刻触发回调
            self.world.enter_tree(child)
            self.world.mark_ready(child)

    def remove_child(self, child: "Node") -> None:
        if child in self.children:
            self.children.remove(child)
        child.parent = None
        if child.in_tree:
            self.world.exit_tree(child)

    def queue_free(self) -> None:
        self.world.pending_free.append(self)

    def free(self) -> None:
        self.world.free_node(self)

    # ---- 回调（由 World 调用） --------------------------------------
    def _enter_tree(self) -> None:
        self.world.log.append(f"{self.name}:enter_tree")

    def _ready(self) -> None:
        self.world.log.append(f"{self.name}:ready")

    def _exit_tree(self) -> None:
        self.world.log.append(f"{self.name}:exit_tree")

    def _process(self, delta: float) -> None:
        self.process_calls += 1
        self.deltas.append(delta)

    def _physics_process(self, delta: float) -> None:
        self.physics_calls += 1
        self.physics_deltas.append(delta)

    def get_node(self, path: str) -> Optional["Node"]:
        if path == "..":
            return self.parent
        if path.startswith("/root"):
            node = self.world.root
            for part in path.split("/")[2:]:
                if not part:
                    continue
                node = next((c for c in node.children if c.name == part), None)
                if node is None:
                    return None
            return node
        return next((c for c in self.children if path in (c.name,)), None)


class PackedScene:
    """场景模板：instantiate() 返回一棵**不在树里**的节点树。"""

    def __init__(self, spec: dict) -> None:
        self.spec = spec

    def instantiate(self, world: "World") -> Node:
        """返回一棵**不在场景树里**的节点树（还要 add_child 才会触发回调）。"""

        def build(name: str, spec: dict) -> Node:
            node = Node(name, world)
            for child_name, child_spec in (spec or {}).items():
                node.children.append(build(child_name, child_spec))
                node.children[-1].parent = node
            return node

        return build(self.spec["name"], self.spec.get("children", {}))


class World:
    def __init__(self) -> None:
        self.root = Node("root", self)
        self.root.in_tree = True
        self.root.ready = True
        self.log: List[str] = []
        self.pending_free: List[Node] = []
        self.physics_acc = 0
        self.elapsed_us = 0

    # ---- 回调派发 ---------------------------------------------------
    def enter_tree(self, node: Node) -> None:
        """_enter_tree：父先于子（自顶向下）。"""
        node.in_tree = True
        node._enter_tree()
        for ch in list(node.children):
            self.enter_tree(ch)

    def mark_ready(self, node: Node) -> None:
        """_ready：子先于父（自底向上）；@onready 在 _ready 之前初始化。"""
        for ch in list(node.children):
            self.mark_ready(ch)
        self.log.append(f"{node.name}:onready")
        node.ready = True
        node._ready()

    def exit_tree(self, node: Node) -> None:
        node.in_tree = False
        node._exit_tree()
        for ch in list(node.children):
            self.exit_tree(ch)

    def free_node(self, node: Node) -> None:
        if not node.alive:
            return
        if node.parent is not None and node in node.parent.children:
            node.parent.children.remove(node)
        if node.in_tree:
            self.exit_tree(node)
        for ch in list(node.children):      # 释放父节点会连带释放所有子节点
            self.free_node(ch)
        node.children.clear()
        node.parent = None
        node.alive = False
        self.log.append(f"{node.name}:free")

    def all_nodes(self, node: Optional[Node] = None) -> List[Node]:
        node = node or self.root
        out = [node]
        for ch in node.children:
            out.extend(self.all_nodes(ch))
        return out

    # ---- 帧循环 ------------------------------------------------------
    def frame(self, dt_us: int) -> None:
        self.elapsed_us += dt_us
        self.physics_acc += dt_us
        phys_steps = 0
        while self.physics_acc >= STEP_US:
            self.physics_acc -= STEP_US
            phys_steps += 1
            for n in self.all_nodes():
                if n.alive and n.in_tree and n.physics_on:
                    n._physics_process(STEP_US / 1_000_000.0)
        for n in self.all_nodes():
            if n.alive and n.in_tree and n.process_on:
                n._process(dt_us / 1_000_000.0)
        for n in list(self.pending_free):   # queue_free：当前帧处理结束后才删
            self.free_node(n)
            self.pending_free.remove(n)


def pacing(total_us: int, frame_us: int) -> List[int]:
    out, left = [], total_us
    while left > 0:
        step = min(frame_us, left)
        out.append(step)
        left -= step
    return out

